### ISSUE TITLE
The input of "Time points of observations"

### ISSUE THREAD

**Issue by chooron** [IanShi1996/IWLatentODE#1]
I read this code and found the model only use the first time for training in the line 151 of 'train.py', which is not suitable for my data, I wonder if time points of each bacth are different, how can I input this time points? (I see other code for latent_ode, they replace the missing data with zero, and provide with a mask matrix, which solve the problem)

**Comment by IanShi1996** [comment]
The quick and easy way is as you describe. I would take the union of all time points in the minibatch, fill non-observed data points with a dummy value, and then use a binary array to denote the positions of each observed data point per trajectory. In the encoder architecture, you could then use the mask array to determine whether to update the hidden state via the GRU or continue to evolve the ODE dynamics. 

Unfortunately I'm not maintaining this repository, but I have attached a code sketch I used for ragged starts and ends which you might be able to adapt for your purposes:

```
        if mask is not None:
            h_arr = torch.zeros(x.size(0), x.size(1), h.size(1)).to(x.device)
            r_fill_mask = self.right_fill_mask(mask)

        for i in range(x.size(1)):
            if i != 0:
                h_ode = odeint(self.node, h, tps[i:i+2], method="euler")[1]

                if mask is not None:
                    curr_rmask = r_fill_mask[:, i].view(-1, 1)
                    h_ode = h_ode * curr_rmask + h * (1 - curr_rmask)
            else:
                # Don't evolve hidden state prior to first observation
                h_ode = h

            h_rnn = self.gru(x[:, i, :], h_ode)

            if mask is not None:
                curr_mask = mask[:, i].view(-1, 1)
                h = h_rnn * curr_mask + h_ode * (1 - curr_mask)

                h_arr[:, i, :] = h
            else:
                h = h_rnn

        if mask is not None:
            ind = self.get_last_tp(mask)
            h = torch.zeros(x.size(0), self.latent_dim * 2).to(x.device)
            for i, traj in enumerate(h_arr):
                h[i] = traj[ind[i]]
```

For the decoder you would also solve for the union of all time points and then pick out which ones are relevant for each trajectory in the minibatch.

**Comment by chooron** [comment]
> 

Thank you very much for your suggestion!  I will utilize the code you provided to modify it later, I hope it can solve my problem.
